Refactor Renderer GUIs

StasJ opened the pull request:
Refactors the GUIs for all renderers, converting them to all use PWidgets. All renderer-related UI files are removed. The subtabs themselves are no longer necessary either and have been removed. PWidgets (along with some new improvements such as a meta "if" PWidget) allow the GUIs to be defined concisely in the EventRouter constructor while everything else is handled automatically. In many cases, the entire GUI definition for a renderer takes less code than previously required just to initialize and update the subtabs which contained the definitions.

Also fixes miscellaneous bugs throughout the application.

clyne commented on apps/vaporgui/AbstractWidgetGroup.h:
```diff
@@ -0,0 +1,39 @@
+#include <initializer_list>
+#include <vector>
+
+template <class This, class Widget>
```
Document. what does this class do?

StasJ replied:
Done

clyne commented on apps/vaporgui/BarbEventRouter.cpp:
```diff
@@ -1,131 +1,45 @@
+            (new PIntegerSliderEdit(BP::_xBarbsCountTag, "X Barbs"))->SetRange(1, 50),
+            (new PIntegerSliderEdit(BP::_yBarbsCountTag, "Y Barbs"))->SetRange(1, 50),
+            (new PIntegerSliderEdit(BP::_zBarbsCountTag, "Z Barbs"))->SetRange(1, 50),
+            (new PDoubleSliderEdit(BP::_lengthScaleTag, "Length Scale"))->SetRange(0.01, 4)->EnableDynamicUpdate(),
```
Does EnableDynamicUpdate enable the user to change the range of the slider? I don't seem to be able to change it. How is it done?

StasJ replied:
`EnableDynamicUpdate()` causes "fast" renders to occur as the user moves the slider for live updates. The slider-edit `VWidget`  itself does not support changing the range.

clyne replied:
By "fast" renders do you  mean the callback isn't called until the slider is released?

Do we have support for changing the range? I thought we did and this is certainly a place where it would be useful.

StasJ replied:
The intermediate updates change the params without adding to the undo history and Render(fast=true) is called. 

The two control range slider I created supports this but VSliderEdit does not support changing the range.

clyne replied:
@stasj Can you add support to VSliderEdit? There are a number of instances where this would be useful.

sgpearse replied:
@clyne @stasj VIntSliderEdit and VDoubleSliderEdit support range modification, FYI.

StasJ replied:
@sgpearse I figured that out eventually. Why do we still have VSliderEdit? It is not documented there that we should be using the other classes.

@clyne done.

clyne replied:
@stasj has enabled user changeable ranges, and will add an option to optionally disable it.

StasJ replied:
Done.

clyne commented on apps/vaporgui/FlowEventRouter.cpp:
```diff
@@ -1,137 +1,144 @@
+    AddSubtab("Seeding", new PGroup({
+        new PSection("Flow Integration Settings", {
+            new PEnumDropdown(FP::_isSteadyTag, {"Streamlines", "Pathlines"}, {true, false}, "Flow Type"),
+            (new Pif(FP::_isSteadyTag))->Equals(true)->Then({
```
Presumably we are only going to use one of these implementations and do away with the other? I.e. before we merge this PR we're going to remove the Pif class use?

StasJ replied:
I don't follow. What are the two implementations?

clyne replied:
I guess I'm, confused. What is Pif and what does the following do?

(new Pif(FP::_isSteadyTag))->Equals(true)->Then({

StasJ replied:
I added some documentation to the Pif class. It may be beneficial to rename it to `PShowIf`
```
(new Pif("param_tag")->Not->Equals("Hi")->Then(
    PWidgets here will only show if param_tag != "Hi"
)->Else(
    PWidgets here will only show if param_tag == "Hi"
);
```

clyne replied:
Got it. Yeah, I'd vote for PShowIf

StasJ replied:
Done.

clyne commented on apps/vaporgui/PAnnotationColorbarWidget.h:
```diff
@@ -0,0 +1,13 @@
+#include "PWidget.h"
+
+class PSection;
+
```
Doxygen documentation please

StasJ replied:
Done

clyne commented on apps/vaporgui/PButton.h:
```diff
@@ -0,0 +1,18 @@
+
+class VPushButton;
+
+class PButton : public PWidget {
```
Doxygen documentation please

StasJ replied:
Done

clyne commented on apps/vaporgui/PCopyRegionWidget.h:
```diff
@@ -0,0 +1,14 @@
+
+class CopyRegionWidget;
+
+class PCopyRegionWidget : public PWidget {
```
Doxygen documentation please

StasJ replied:
Done

clyne commented on apps/vaporgui/PGroup.h:
```diff
@@ -1,25 +1,27 @@
 //! \copydoc PWidget
 
-class PGroup : public PWidget {
+class PGroup : public PWidget, public WidgetGroupWrapper<PGroup, PWidget, VGroup> {
```
Doxygen documentation please

StasJ replied:
This already has Doxygen documentation.

clyne commented on apps/vaporgui/POrientationSelector.h:
```diff
@@ -0,0 +1,13 @@
+
+class PEnumDropdown;
+
+class POrientationSelector : public PWidget {
```
Doxygen documentation please

StasJ replied:
Done

clyne commented on apps/vaporgui/SliceEventRouter.cpp:
```diff
@@ -1,140 +1,33 @@
+        new PSection("Slice", {
+            new POrientationSelector,
+            new PSliceSampleLocationSelector,
+            (new PDoubleSliderEdit(SliceParams::_sampleRateTag, "N Samples"))->SetRange(32, 2000)
```
The current version of the GUI doesn't explicitly expose the number of samples to use to the user. That is an implementation detail. Instead the user is presented with a unitless quality control knob. Under the hood, higher quality values result in more samples. I'm curious as to what @sgpearse and @shaomeng think

clyne commented on apps/vaporgui/Updateable.h:
```diff
@@ -0,0 +1,12 @@
+    class ParamsMgr;
+    class DataMgr;
+}
+
```
Doxygen document please

StasJ replied:
Done

clyne commented on apps/vaporgui/V3DInput.h:
```diff
@@ -0,0 +1,23 @@
+
+class VDoubleLineEdit;
+
+class V3DInput : public QWidget {
```
Doxygen

StasJ replied:
Done

clyne commented on apps/vaporgui/VSection.h:
```diff
@@ -38,6 +38,20 @@ class VSection : public QTabWidget {
+#include "AbstractWidgetGroup.h"
+#include "VGroup.h"
+
+class VSectionGroup : public VSection, public WidgetGroupWrapper<VSectionGroup, QWidget, VGroup> {
```
Doxygen

StasJ replied:
Done

clyne commented on include/vapor/BarbParams.h:
```diff
@@ -42,17 +42,17 @@ class PARAMS_API BarbParams : public RenderParams {
-	defaultGrid.push_back(1.0);
-	return (GetValueLongVec(_gridTag,defaultGrid));
+     vector<long> dims(3);
+     dims[0] = GetValueLong(_xBarbsCountTag, 10);
```
changing the parameter tags will make older session files incompatible, will it not? It would be good to avoid this if we can, but we can discuss.

StasJ replied:
In the sense that certain renderers may lose a part of their saved state but the session files will still be 95% ok. It is possible to achieve similar functionality without making these changes. We can discuss pros/cons at our meeting.

clyne commented on include/vapor/ColorbarPbase.h:
```diff
@@ -146,11 +146,13 @@ class PARAMS_API ColorbarPbase : public ParamsBase {
 	static const string _colorbarBackColorTag;
-	static const string _colorbarSizeTag;
-	static const string _colorbarPositionTag;
+	static const string _colorbarSizeXTag;
```
Will this break compatibility?

StasJ replied:
In the sense that certain renderers may lose a part of their saved state but the session files will still be 95% ok. It is possible to achieve similar functionality without making these changes. We can discuss pros/cons at our meeting.

clyne commented on lib/params/RenderParams.cpp:
```diff
@@ -490,10 +497,10 @@ void RenderParams::SetFieldVariableNames(vector<string> varnames){
-		_fieldVariableNamesTag, "Specify vector field variable names", 
-		varnames
-	);
+//	SetValueStringVec(_fieldVariableNamesTag, "", varnames);
```
can we remove commented out code

StasJ replied:
Done

clyne commented on lib/params/SliceParams.cpp:
```diff
@@ -69,10 +70,17 @@ int SliceParams::Initialize() {
-    minExt[Z] = average;
-    maxExt[Z] = average;
-    box->SetExtents(minExt, maxExt);
+//    double average = (minExt[Z] + maxExt[Z])/2.f;
```
Commented out code needed?

StasJ replied:
Removed.

clyne reviewed (commented):
Really nice work on this, Stas! The code has been tremendously simplified, and should present the user with a much more consistent UI. My only concern is that there appear to have been changes made to the Params classes that will make earlier session files incompatible with this change. True? If so, we should all discuss.

StasJ commented on apps/vaporgui/BarbEventRouter.cpp:
```diff
@@ -1,131 +1,45 @@
-	);
+    AddSubtab("Variables", new PGroup({
+        new PSection("Variable Selection", {
+            new S::PDimensionSelector,
```
S is just a temporary namespace which houses the original PWidget variable selectors since the ones in master are currently broken.

clyne replied:
I couldn't find a definition for it. Will it be replaced before this PR is merged? Can't the original PWidge variable selectors be replaced by this PR?

StasJ replied:
They are defined in `S_PVariableSelector.h` and `S_PDimensionSelector.h`. 

Im fine with either solution. The original widgets can either be fixed in which case we can just remove the S:: namespace or we can replace the original widgets the ones in this PR.

clyne replied:
I don't have a preference of one of widgets over the other. If one is working (and is fully functional), let's use it.

clyne replied:
@sgpearse please review and indicate whether we need to fix the old widgets or we can simply use the new, working implementation.

sgpearse replied:
@clyne seems to work well for me with the new eventFilter code, so I don't see any reason to not adopt this.  I just have a few questions and comments in my review.

clyne replied:
@StasJ let's do it

StasJ replied:
@clyne Done.

shaomeng reviewed (approved):
This PR seems to be pretty comprehensive and carefully designed. Good job Stas!

clyne commented on apps/vaporgui/BarbEventRouter.cpp:
```diff
@@ -1,131 +1,45 @@
+            (new PIntegerSliderEdit(BP::_yBarbsCountTag, "Y Barbs"))->SetRange(1, 50),
+            (new PIntegerSliderEdit(BP::_zBarbsCountTag, "Z Barbs"))->SetRange(1, 50),
+            (new PDoubleSliderEdit(BP::_lengthScaleTag, "Length Scale"))->SetRange(0.01, 4)->EnableDynamicUpdate(),
+            (new PDoubleSliderEdit(BP::_thicknessScaleTag, "Thickness Scale"))->SetRange(0.01, 4)->EnableDynamicUpdate(),
```
I see no reason to remove API simply because the current UI doesn't use it. A future one (e.g. Python, web, etc.) might.

sgpearse replied:
@clyne 

The other interfaces should also use the LLI in this circumstance, since there is no validation being performed.

clyne replied:
I disagree. It is not unreasonable for APIs to provide multiple ways to accomplish the same thing.

sgpearse commented on apps/vaporgui/SliceEventRouter.cpp:
```diff
@@ -1,140 +1,33 @@
+    AddSubtab("Appearance", new PGroup({
+        new PTFEditor,
+        new PSection("Slice", {
+            new POrientationSelector,
```
The orientation should go in the Geometry tab.

StasJ replied:
I thought it was more intuitive to put it here but we can discuss at our meeting.

sgpearse commented on apps/vaporgui/SliceEventRouter.cpp:
```diff
@@ -1,140 +1,33 @@
+        new PTFEditor,
+        new PSection("Slice", {
+            new POrientationSelector,
+            new PSliceSampleLocationSelector,
```
This also should be in the Geometry tab.  Could you change the label to be something like "Z Plane Location" or give it a tool tip?  "Sample Z" wasn't clear to me off the bat.

StasJ replied:
I thought it was more intuitive to put it here but we can discuss at our meeting.

sgpearse commented on apps/vaporgui/PSliceSampleLocationSelector.h:
```diff
@@ -0,0 +1,37 @@
+    PSliceSampleLocationSelector();
+};
+
+class PSliceSampleLocationSelector1D : public PLineItem {
```
Should this class be using a VDoubleSliderEdit?

StasJ replied:
I guess so. I've replace it. Why do we still have `VSliderEdit` then?

sgpearse replied:
It just hasn't been replaced throughout the application.

sgpearse commented on apps/vaporgui/PButton.cpp:
```diff
@@ -0,0 +1,13 @@
+#include "VPushButton.h"
+
+PButton::PButton(std::string label, Callback cb)
+: PWidget("", _button = new VPushButton(label)), _cb(cb)
```
This seems to be a new form of PWidget that uses a callback, not the HLI, or (necessarily) the LLI.  Can you talk about what was insufficient about the LLI/HLI method we use everywhere else?

StasJ replied:
The LLI and HLI are used to sync a widget with the params database. Given that this is a button, it is not synced with the params database so it does not make sense to use those functions.

sgpearse commented on apps/vaporgui/FlowEventRouter.cpp:
```diff
@@ -1,137 +1,144 @@
+            new PDoubleInput(FP::_velocityMultiplierTag, "Vector Field Multiplier"),
+            new PCheckbox(FP::_xPeriodicTag, "X axis periodicity"),
+            new PCheckbox(FP::_yPeriodicTag, "Y axis periodicity"),
+            new PCheckbox(FP::_zPeriodicTag, "Z axis periodicity"),
```
I believe we are now setting default values in the Params constructors.  Correct?

StasJ replied:
PWidgets try to avoid any notion of "defaults". Since they are required for the params API, all numbers are defaulted to 0 and all strings are empty.

Correct.

StasJ commented on apps/vaporgui/AbstractWidgetGroup.cpp:
```diff
@@ -0,0 +1 @@
+#include "AbstractWidgetGroup.h"
```
It ensures that the header is always compiled. It is also standard for a class to have both a header and implementation file.

shaomeng replied:
Can you explain why simply including a header file can make it compiled?

StasJ commented on apps/vaporgui/ContourEventRouter.h:
```diff
@@ -1,92 +1,30 @@
-	ContourAppearanceSubtab *_appearance;
-	ContourAnnotationSubtab *_annotation;
+    PSliderEdit *_spacingSlider;
+    PSliderEdit *_minValueSlider;
```
I didn't see any reason to declare them as HLI widgets.

StasJ replied:
It is an implementation detail that is not relevant to the interface.

StasJ commented on apps/vaporgui/PWidgetWrapper.h:
```diff
@@ -0,0 +1,12 @@
+
+#include "PWidget.h"
+
+class PWidgetWrapper : public PWidget {
```
Done.

StasJ commented on apps/vaporgui/PSection.cpp:
```diff
@@ -3,11 +3,12 @@
 {
     _pgroup = new PGroup;
     _vsection->layout()->addWidget(_pgroup);
+    _pgroup->AddM(widgets);
```
`AddM` is just an overloaded function of `Add` but it has to have a different name due to a quirk in how c++ handles virtual functions.

StasJ commented on apps/vaporgui/VHBoxWidget.h:
```diff
@@ -0,0 +1,44 @@
+    Q_OBJECT
+
+public:
+//    virtual QSize sizeHint() const;
```
It broke the GUI layout so I had to remove it.

sgpearse replied:
Hmm.  I tried re-adding it and it seemed to work to me.  Do you remember how it broke?  I'd like to keep this since I was asked to add it, but if it really is broken I can revisit it.

clyne reviewed (approved):
Great work, Stas!

clyne commented:
@StasJ I approved this PR and was about to merge it, but it looks like there are some conflicts that need to be resolved :-(

StasJ commented:
@clyne I rebased off of the most recent master and fixed the conflicts however the branch will no longer compile because the GeoImageTMS PR introduced a circular dependency. Image params uses GeoImageTMS which is a part of the renderer library which depends on the params library. @sgpearse can you please resolve this?
